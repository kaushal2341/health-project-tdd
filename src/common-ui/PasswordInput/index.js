import React from 'react';
import Proptypes from 'prop-types';
import {Icon} from 'semantic-ui-react'
class PasswordInput extends React.PureComponent{
    state={
          errorMsg:"",
          showPassword:true
          };
    setTheState = msg =>{
        this.setState({errorMsg:msg});
    }
    checkValidLength = value =>{
        let errorMsg="";
        let valueToBeReturned=false
        if(value.length > 5)
        valueToBeReturned = true; 
        else if(value.length === 0)
        errorMsg=""
        else
        errorMsg="Sorry the length Of password must be greater than 5"
        this.setTheState(errorMsg);
        return valueToBeReturned; 
    }
    checkPasswordPolicy = (value , check) => {
        let errorValueToBeSet = "";
        if(check)
        { 
        const weakPasswordPolicy=/^[a-zA-Z]+$/;
        const strongPasswordPolicy = new 
                RegExp("^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&\*])(?=.{8,})");
          
        errorValueToBeSet = value.match(weakPasswordPolicy) 
                            ? "WEAK"
                            : value.match(strongPasswordPolicy)
                            ? "STRONG"
                            : "MEDIUM"    
        }
        if(errorValueToBeSet)
           this.setTheState(errorValueToBeSet);
    }
    onChangeHandler = (e)=>{
      let shouldCall=true;
      const {value} = e.target
      shouldCall=this.checkValidLength(value);
      this.checkPasswordPolicy(value,shouldCall);
      if(shouldCall)
      this.props.onChangeHandler(e);
    }
    onSeePassword = (e) => {
     this.setState((prevState)=>({
       showPassword:!prevState.showPassword 
     }))
    }
    render(){
     return(<>
            <input 
            id={this.props.id}
            name={this.props.name}
            type={this.state.showPassword ? "password" : "text"}
            minLength={this.props.min}
            maxLength={this.props.max}
            onChange={this.onChangeHandler}
            onBlur={this.props.onBlur}
            onFocus={this.props.onFocus}
            className={this.props.classes}
            placeholder={this.props.placeholder}
            required={this.props.required}
            readOnly={this.props.readonly}
            />
            <button className="btn btn-sm" onClick={this.onSeePassword}><Icon className={!this.state.showPassword ? 'eye' : 'eye slash'}/></button>
            <span id="error">{this.state.errorMsg}</span>
            </>
            )
 } 
}
React.propTypes ={
    id:Proptypes.string||Proptypes.number,
    name:Proptypes.string.isRequired,
    min:Proptypes.number,
    max:Proptypes.number,
    onChange:Proptypes.func.isRequired,
    onBlur:Proptypes.func,
    onFocus:Proptypes.func,
    classes:Proptypes.string,
    placeholder:Proptypes.string,
    required:Proptypes.bool,
    readOnly:Proptypes.bool
}
export default PasswordInput;