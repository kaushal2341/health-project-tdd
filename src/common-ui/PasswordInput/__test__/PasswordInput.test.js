import React from 'react';
import PasswordInput from '..';
import {Input} from 'semantic-ui-react';
expect.addSnapshotSerializer(enzymeSerializer);
describe('Password Input',()=>{
let wrapper;
let onChangeMockFunc;
beforeEach(()=>{
onChangeMockFunc = jest.fn()
wrapper = mount(<PasswordInput onChangeHandler={onChangeMockFunc}/>);
});
afterEach(()=>{
    wrapper.unmount()
})
it('should be defined',()=>{
   expect(wrapper).toBeDefined();
});
it('should have only one input tag element',()=>{
    expect(wrapper.find('input')).toHaveLength(1);
});
it('should have password input tag element',()=>{
    expect(wrapper.find('input').prop('type')).toBe('password');
});
it('should have the given props in input tag element',()=>{
    const reqProps=['id',
                    'name',
                    'type',
                    'minLength',
                    'maxLength',
                    'onChange',
                    'onBlur',
                    'onFocus',
                    'className',
                    'placeholder',
                    'required',
                    'readOnly',
                    'children',
                    'error',
                    'fluid',
                    'focus',
                    'icon',
                    'iconPosition',
                    'inverted',
                    'label',
                    'labelPosition',
                    'loading',
                    'size',
                    'transparent',
                    'tabIndex',
                    'action',
                    'actionPosition',
                    'disabled',
                    'value'
                  ]
    let passwordInputPropsKeys=Object.keys(wrapper.find(Input).props());
    reqProps.map((reqProp,i)=>{
        expect(reqProps).toContain(passwordInputPropsKeys[i]);
    })
});
it('should have one span tag element to show error msg',()=>{
    expect(wrapper.find('span')).toHaveLength(1);
})
describe('Password Input Simulation',()=>{
    function simulationChange(value){
        wrapper.find('input').simulate('change',value);
    }
    it('should call the onChange function when input is changed',()=>{
        simulationChange({target:{value:'sdsdsd'}})
        expect(onChangeMockFunc).toHaveBeenCalledTimes(1); 
    });
    
    it('should have state of errorMsg ',()=>{
        expect(Object.keys(wrapper.state())).toContain('errorMsg');
    });

    it('should show the error message when the length is less than 6',()=>{
        simulationChange({target:{value:'abcd'}});
        expect(wrapper.state('errorMsg')).toEqual('Sorry the length Of password must be greater than 5');
    });

    it('should not show the error message when the length is less than 1',()=>{
        simulationChange({target:{value:''}});
        expect(wrapper.state('errorMsg')).toEqual('');
    });

    it('should show WEAK error message when the value match weak regexPolicy',()=>{
        simulationChange({target:{value:'abcdef'}});
        expect(wrapper.state('errorMsg')).toEqual('WEAK');
    });
    it('should show STRONG error message when the value match strong regexPolicy',()=>{
        simulationChange({target:{value:'abcdefABCDEF123@@#'}});
        expect(wrapper.state('errorMsg')).toEqual('STRONG');
    });
    it('should show MEDIUM error message when the value match strong regexPolicy',()=>{
        simulationChange({target:{value:'abcdef1'}});
        expect(wrapper.state('errorMsg')).toEqual('MEDIUM');
    });
    it('should have state of showPassword ',()=>{
        expect(Object.keys(wrapper.state())).toContain('showPassword');
    });
    it(' should have only one i tag', ()=>{
        expect(wrapper.find('i')).toHaveLength(1);
    });
    
    it('should have onClick prop on tag i', () => {
      expect(wrapper.find('i').prop('onClick')).toBeDefined()
    });
    
    it('should change showPassword value when clicked on tag i',()=>{
      wrapper.find('i').simulate('click');
      expect(wrapper.state('showPassword')).toBeFalsy();    
    });
    
});
   describe('Snapshot Testing',()=>{
        it('should create or match previous snapshot of the component',() => {
            expect(wrapper).toMatchSnapshot(); 
        });
    });
   
})