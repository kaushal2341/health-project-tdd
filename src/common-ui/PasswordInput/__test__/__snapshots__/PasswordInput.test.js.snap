// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Input Snapshot Testing should create or match previous snapshot of the component 1`] = `
<PasswordInput
  onChangeHandler={[MockFunction]}
>
  <Input
    onChange={[Function]}
    type="password"
  >
    <div
      className="ui input"
    >
      <input
        onChange={[Function]}
        type="password"
      />
    </div>
  </Input>
  Â 
  <ButtonInput
    onClickHandler={[Function]}
    size="tiny"
  >
    <Button
      as="button"
      onClick={[Function]}
      size="tiny"
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui tiny button"
            >
              <i
                aria-hidden="true"
                class="icon eye slash"
              />
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui tiny button"
              >
                <i
                  aria-hidden="true"
                  class="icon eye slash"
                />
              </button>,
            }
          }
        >
          <button
            className="ui tiny button"
            onClick={[Function]}
          >
            <Icon
              as="i"
              className="eye slash"
            >
              <i
                aria-hidden="true"
                className="icon eye slash"
                onClick={[Function]}
              />
            </Icon>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
  </ButtonInput>
  <span
    id="error"
  />
</PasswordInput>
`;
